homeassistant/components/heos/__init__.py

```diff
@@ -219,6 +228,125 @@ def update_ids(self, mapped_ids: dict[int, int]):
+            return group_info_by_entity_id
+
+        player_id_to_entity_id_map = self._hass.data[DOMAIN][DATA_ENTITY_ID_MAP]
+        for group in groups.values():
+            leader_entity_id = player_id_to_entity_id_map.get(group.leader.player_id)
+            member_entity_ids = [
+                player_id_to_entity_id_map[member.player_id]
+                for member in group.members
+                if member.player_id in player_id_to_entity_id_map
+            ]
+            # Make sure the group leader is always the first element
+            group_info = [leader_entity_id, *member_entity_ids]
+            if leader_entity_id:
+                group_info_by_entity_id[leader_entity_id] = group_info
+                for member_entity_id in member_entity_ids:
+                    group_info_by_entity_id[member_entity_id] = group_info
+
+        return group_info_by_entity_id
+
+    async def async_join_players(
+        self, leader_entity_id: str, member_entity_ids: list[str]
+    ) -> None:
+        """Create a group with `leader_entity_id` as group leader and `member_entity_ids` as member players."""
+        entity_id_to_player_id_map = self._get_entity_id_to_player_id_map()
+        leader_id = entity_id_to_player_id_map.get(leader_entity_id)
+        member_ids = [
+            entity_id_to_player_id_map[member]
+            for member in member_entity_ids
+            if member in entity_id_to_player_id_map
+        ]
+
+        if leader_id:
+            try:
+                await self.controller.create_group(leader_id, member_ids)
+            except HeosError as err:
+                _LOGGER.error(
+                    "Failed to group %s with %s: %s",
+                    leader_entity_id,
+                    member_entity_ids,
+                    err,
+                )
+
+    async def async_unjoin_player(self, player_entity_id: str):
+        """Remove `player_entity_id` from any group."""
+        player_id = self._get_entity_id_to_player_id_map().get(player_entity_id)
+
+        if player_id:
+            try:
+                await self.controller.create_group(player_id, [])
+            except HeosError as err:
+                _LOGGER.error(
+                    "Failed to ungroup %s: %s",
+                    player_entity_id,
+                    err,
+                )
+
+    async def async_update_groups(self, event, data=None):
+        """Update the group membership from the controller."""
+        if event in (
+            heos_const.EVENT_GROUPS_CHANGED,
+            heos_const.EVENT_CONNECTED,
+            SIGNAL_HEOS_PLAYER_ADDED,
+        ):
+            groups = await self.async_get_group_membership()
+            if groups:
+                self._group_membership = groups
+                _LOGGER.debug("Groups updated due to change event")
+                # Let players know to update
+                self._hass.helpers.dispatcher.async_dispatcher_send(SIGNAL_HEOS_UPDATED)
```

Comment:
Please import `async_dispatcher_send` instead. We prefer to import to make typing work.

Reply:
Thanks for pointing that out. I am importing `async_dispatcher_connect`/`async_dispatcher_send` now.